### PR TITLE
Avoid a scrape_region_constraints and instead register the region constraints directly

### PR DESCRIPTION
**Pull request by oli-obk**
Should fix the perf regression from https://github.com/rust-lang/rust/pull/122077#issuecomment-2044027926

### CODE REVIEW

**Comment by oli-obk** [comment]
yay

**Comment by compiler-errors** [comment]
How does this fix the original issue that this scrape was added for? Didn't a bunch of crates ICE in crater? Did we ever minimize them down to UI tests?

**Comment by compiler-errors** [comment]
Oh, I see that it's equating the opaque manually... I wonder if that's sound wrt generalization.

File: compiler/rustc_infer/src/infer/canonical/query_response.rs [hunk 1]
```diff
@@ -248,64 +248,71 @@ impl<'tcx> InferCtxt<'tcx> {
...
+            // variable that got constrained to the opaque type itself. In that case we want to ensure
+            // any lifetime differences get recorded in `ouput_query_region_constraints` instead of
+            // being registered in the `InferCtxt`.
+            match hidden_ty.kind() {
+                ty::Alias(ty::Opaque, alias_ty)
+                    if alias_ty.def_id == opaque_type_key.def_id.into() =>
+                {
+                    assert_eq!(alias_ty.args.len(), opaque_type_key.args.len());
+                    for (key_arg, hidden_arg) in
+                        opaque_type_key.args.iter().zip(alias_ty.args.iter())
```
Comment (compiler-errors) [inline compiler/rustc_infer/src/infer/canonical/query_response.rs:271]: This isnt taking into account variance is it?

Comment (oli-obk) [reply compiler/rustc_infer/src/infer/canonical/query_response.rs:271]: Nope, but that shouldn't matter here in practice, as all the args are either params or inference vars. Well, they should be. Unfortunately in the presence of erroneous code we may not have errored yet, as that only happens at the end of borrowck.

**Comment by oli-obk** [comment]
> Did we ever minimize them down to UI tests?

Yes, and I confirmed it registers lifetimes and fails in borrowck with the ICE we saw on crater

**Comment by apiraino** [comment]
To all: is this waiting on #120700 or other design? thanks

**Comment by lcnr** [comment]
The perf impact has been meaningful, I am however fairly uncomfortable with merging this PR as stated above.

With the new solver we'll likely want to perform pretty much all type-ops locally again as the new trait solver has a far better global cache. In this case keeping the outer `scrape_region_constraints` has a far smaller impact.

If we were to go and lean into this `scrape_region_constraints` we should remove `type_op_eq` and `type_op_subtype` and always do them locally, even with the old solver. Canonicalizing just to equate types very much does not feel worth it to me

**Pull request closed without merging**